components/raftstore-v2/src/operation/ready/snapshot.rs

```diff
@@ -0,0 +1,290 @@
+    // index used to check if the gen task should be canceled.
+    index: Arc<AtomicU64>,
+    // Set it to true to cancel the task if necessary.
+    canceled: Arc<AtomicBool>,
+    // indicates whether the snapshot is triggered due to load balance
+    for_balance: bool,
+}
+
+impl GenSnapTask {
+    pub fn new(region_id: u64, index: Arc<AtomicU64>, canceled: Arc<AtomicBool>) -> GenSnapTask {
+        GenSnapTask {
+            region_id,
+            index,
+            canceled,
+            for_balance: false,
+        }
+    }
+
+    pub fn set_for_balance(&mut self) {
+        self.for_balance = true;
+    }
+}
+
+impl Debug for GenSnapTask {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.debug_struct("GenSnapTask")
+            .field("region_id", &self.region_id)
+            .finish()
+    }
+}
+
+impl<EK: KvEngine, ER: RaftEngine> Peer<EK, ER> {
+    pub fn on_snapshot_generated(&mut self, snapshot: Box<Snapshot>) {
+        if self.storage_mut().on_snapshot_generated(snapshot) {
+            self.raft_group_mut().ping();
+            self.set_has_ready();
+        }
+    }
+}
+
+impl<EK: KvEngine, R: ApplyResReporter> Apply<EK, R> {
+    /// Handle snapshot.
+    ///
+    /// Will schedule a task to read worker and then generate a snapshot
+    /// asynchronously.
+    pub fn schedule_gen_snapshot(&mut self, snap_task: GenSnapTask) {
+        // Flush before do snapshot.
+        if snap_task.canceled.load(Ordering::SeqCst) {
+            return;
+        }
+        self.flush();
+
+        // Send generate snapshot task to region worker.
+        let (last_applied_index, last_applied_term) = self.apply_progress();
+        snap_task.index.store(last_applied_index, Ordering::SeqCst);
+        let gen_tablet_sanp_task = ReadTask::GenTabletSnapshot {
+            region_id: snap_task.region_id,
+            tablet: self.tablet().clone(),
+            region_state: self.region_state().clone(),
+            last_applied_term,
+            last_applied_index,
+            for_balance: snap_task.for_balance,
+            canceled: snap_task.canceled.clone(),
+        };
+        if let Err(e) = self.read_scheduler().schedule(gen_tablet_sanp_task) {
+            error!(
+                self.logger,
+                "schedule snapshot failed";
+                "error" => ?e,
+            );
+            snap_task.canceled.store(true, Ordering::SeqCst);
+        }
+    }
+}
+
+impl<EK: KvEngine, ER: RaftEngine> Storage<EK, ER> {
+    /// Gets a snapshot. Returns `SnapshotTemporarilyUnavailable` if there is no
+    /// unavailable snapshot.
+    pub fn snapshot(&self, request_index: u64, to: u64) -> raft::Result<Snapshot> {
+        let mut snap_state = self.snap_state_mut();
+        match *snap_state {
+            SnapState::Generating { ref canceled, .. } => {
+                if canceled.load(Ordering::SeqCst) {
+                    self.cancel_generating_snap(None);
```

Comment:
Seems can be simplified as `*snap_state = SnapState::Relax; metrics.inc()`

Reply:
I think it's better unify use cancel function here, it's not in key path not hurt performance.